### PR TITLE
Initial implementation for clusterstatus resource

### PR DESCRIPTION
**Pull request by tuommaki**
Add initial implementation for clusterstatus resource which covers provider
independent parts of cluster status management. Most notable responsibilities
of this code are:

  - Cluster condition management (i.e. Creating, Created, Updating, Updated, etc.).
  - Cluster version history management.

Note that `Deleting` conditions need to be addressed still. We do this in the next PRs as we are so far only concerned with cluster creation.

### CODE REVIEW

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go [hunk 1]
```diff
@@ -2,8 +2,242 @@ package clusterstatus
...
+	cr.Labels[label.Cluster] = updatedStatus.ID
+
+	var nodes []corev1.Node
+	var tenantAPIAvailable bool
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding nodes of tenant cluster")
+
+		l, err := cc.Client.TenantCluster.K8s.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenantcluster.IsTimeout(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "timeout fetching certificates")
```
Comment (TheoBrigitte) [inline service/controller/clusterapi/v17/resources/clusterstatus/create.go]: You should return the error after logging.

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go [hunk 2]
```diff
@@ -2,8 +2,242 @@ package clusterstatus
...
+		} else if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant API not available")
```
Comment (TheoBrigitte) [inline service/controller/clusterapi/v17/resources/clusterstatus/create.go]: Here also

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go [hunk 3]
```diff
@@ -2,8 +2,242 @@ package clusterstatus
...
+	if clusterID != "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found out cluster has cluster label containing %s; reusing that", status.ID))
+		status.ID = clusterID
+		return status, nil
+	}
+
+	return status, microerror.Maskf(notFoundError, "cluster ID")
+}
+
+func (r *Resource) computeClusterConditions(ctx context.Context, cluster cmav1alpha1.Cluster, clusterStatus v1alpha1.CommonClusterStatus, tenantAPIAvailable bool, nodes []corev1.Node, machineDeployments []cmav1alpha1.MachineDeployment) v1alpha1.CommonClusterStatus {
```
Comment (TheoBrigitte) [inline service/controller/clusterapi/v17/resources/clusterstatus/create.go]: I would like to have a unit test for this function or re-write the conditions in a way that guarantee there is only one action taken.

The different block logic here make reasoning about the conditions hard I feel. I would prefer to gather in the first places all the elements needed to take a decision and have one if/else block in the end which show how the decision is taken instead of having those split up into different blocks.

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: We should not log and return errors. This is an anti pattern as the same information gets managed twice. When you operate a service this is pretty confusing IMO. The error here is expected and we do it like this in a lot of cases. This is fine.

Comment (tuommaki) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: Yes, both of these two cases are such that overall system behaviour would be in dead lock if there was return here. When cluster is created, it's expected that there are no certs (because this resource gets executed before Cluster -> AWSClusterConfig mapping resource and certs are created from AWSClusterConfig) or that API is down (because there were no secrets, certs or even cluster ID in status so that aws-operator could create the cluster).

TL;DR; As Tim already commented above, this is fine. :slightly_smiling_face:

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go (hunk 2 above)
Comment (tuommaki) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: See above. :slightly_smiling_face:

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go (hunk 3 above)
Comment (tuommaki) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: After getting this more or less ready, I was thinking the same. I'll see how I can change this so that it's a bit clearer. Initially I was about to add returns on conditions, but version management below is required occasionally and return would break that. I'll come back to you a bit later on this. :+1:

**Comment by xh3b4sd** [comment]
Taking over here now.

**Comment by xh3b4sd** [comment]
@giantswarm/sig-operator please have another look here.

**Comment by xh3b4sd** [comment]
Fishing for LGTM.

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go [hunk 4]
```diff
@@ -2,8 +2,212 @@ package clusterstatus
...
+		return nil
+	}
+
+	var nodes []corev1.Node
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding nodes of tenant cluster")
+
+		l, err := cc.Client.TenantCluster.K8s.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenant.IsAPINotAvailable(err) {
+			// Similarly to above timeout, during cluster creation it is
```
Comment (tuommaki) [inline service/controller/clusterapi/v17/resources/clusterstatus/create.go]: This comment lags behind. It should probably be something like this:
```
During cluster creation / upgrade the tenant API is naturally not available but this resource must still continue execution as that's when `Creating` and `Upgrading` conditions happen.
```

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go [hunk 5]
```diff
@@ -2,8 +2,212 @@ package clusterstatus
...
+	{
+		notCreating := !clusterStatus.HasCreatingCondition()
+		conditionsEmpty := len(clusterStatus.Conditions) == 0
+		versionsEmpty := len(clusterStatus.Versions) == 0
+
+		if notCreating && conditionsEmpty && versionsEmpty {
+			clusterStatus.Conditions = clusterStatus.WithCreatingCondition()
+			r.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("setting %#q status condition", v1alpha1.ClusterStatusConditionCreating))
+		}
+	}
```
Comment (tuommaki) [inline service/controller/clusterapi/v17/resources/clusterstatus/create.go]: These are still fall-through-style as commented by Theo. I can live with these for now as this is how the earlier statusresource was as well. I think this is something that should be addressed as separate PR.

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go (hunk 4 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: Done

File: service/controller/clusterapi/v17/resources/clusterstatus/create.go (hunk 5 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v17/resources/clusterstatus/create.go]: As discussed. I agree with the unit tests. I do not agree on the different style. Will provide tests in a separate PR. This code was more or less copied from `statusresource` and works like that in production for almost a year already.
